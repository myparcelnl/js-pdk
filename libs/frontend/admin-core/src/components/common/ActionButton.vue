<template>
  <PdkButton
    v-test="`actionButton--${action?.id}`"
    :aria-label="hideText ? action?.label : null"
    :disabled="disabled || action?.disabled"
    :icon="action?.icon"
    :label="!hideText ? action?.label : null"
    :loading="loading"
    :size="size"
    :title="hideText ? action?.label : null"
    :variant="variant"
    @click="onClick" />
</template>

<script lang="ts" setup>
import {Size, Variant} from '@myparcel-pdk/common';
import {ActionDefinition} from '../../types';
import {PropType} from 'vue';
import {useActionStore} from '../../stores';
import {useLoading} from '../../composables';

const props = defineProps({
  action: {
    type: Object as PropType<ActionDefinition>,
    required: true,
  },

  disabled: {
    type: Boolean,
  },

  hideText: {
    type: Boolean,
  },

  size: {
    type: String as PropType<Size>,
    default: Size.Medium,
  },

  variant: {
    type: String as PropType<Variant>,
    default: Variant.Primary,
  },
});

const emit = defineEmits(['click']);

const {loading, setLoading} = useLoading();

const actionStore = useActionStore();

const onClick = async () => {
  emit('click');
  setLoading(true);
  await actionStore.dispatch(props.action.id, props.action.parameters);
  setLoading(false);
};
</script>
