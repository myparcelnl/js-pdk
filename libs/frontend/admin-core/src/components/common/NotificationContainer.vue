<template>
  <div v-test="`NotificationContainer--${category}`">
    <TransitionGroup
      :name="config?.transitions?.notification"
      appear>
      <PdkNotification
        v-for="(notification, index) in notifications"
        :key="`alert_${index}_${notification.content}`"
        :notification="notification" />
    </TransitionGroup>
  </div>
</template>

<script lang="ts">
import {PropType, computed, defineComponent} from 'vue';
import {NotificationCategory} from '../../types';
import {useNotificationStore} from '../../stores';
import {useAdminConfig} from '../../composables';

export default defineComponent({
  name: 'NotificationContainer',

  props: {
    category: {
      type: String as PropType<NotificationCategory>,
      default: () => NotificationCategory.General,
    },
  },

  setup: (props) => {
    return {
      config: useAdminConfig(),

      notifications: computed(() => {
        const {notifications} = useNotificationStore();

        return notifications.filter((notification) => {
          return notification.category === props.category;
        });
      }),
    };
  },
});
</script>
