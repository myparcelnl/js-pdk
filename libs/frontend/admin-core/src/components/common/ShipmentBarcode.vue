<template>
  <span
    v-test="`ShipmentBarcode--${shipment.id}`"
    :class="[config?.cssUtilities?.whitespaceNoWrap, config?.cssUtilities?.displayFlex]">
    <PdkImage
      v-if="carrier"
      :alt="carrier?.human"
      :src="useAssetUrl(carrier?.meta.logo_svg)"
      :title="carrier?.human"
      width="20" />

    <template v-if="carrier">&nbsp;</template>

    <PdkLink
      v-if="shipment.barcode && shipment.linkConsumerPortal"
      :alt="translate('shipment_barcode')"
      :href="shipment.linkConsumerPortal">
      {{ shipment.barcode }}
    </PdkLink>

    <span
      v-else
      :title="translate('shipment_barcode')"
      v-text="shipment.barcode || translate('shipment_no_barcode')"></span>
  </span>
</template>

<script lang="ts" setup>
import {useQueryStore} from '../../stores';
import {useAdminConfig, useLanguage, useShipmentData} from '../../composables';

const props = defineProps<{shipmentId: number}>();

useQueryStore().registerShipmentQuery(props.shipmentId);

const {translate} = useLanguage();

const {useAssetUrl, carrier, shipment} = useShipmentData(props.shipmentId);
const config = useAdminConfig();
</script>
