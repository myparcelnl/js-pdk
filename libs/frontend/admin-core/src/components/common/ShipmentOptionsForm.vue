<template>
  <MagicForm :form="shipmentOptionsForm" />
</template>

<script setup lang="ts">
import {MagicForm} from '@myparcel/vue-form-builder/src';
import {createShipmentOptionsForm} from '../../forms';
import {get} from '@vueuse/core';
import {useOrders} from '../../composables';

const queries = useOrders();

// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
const shipmentOptionsForm = createShipmentOptionsForm(queries.map((query) => get(query.data)!));
</script>
