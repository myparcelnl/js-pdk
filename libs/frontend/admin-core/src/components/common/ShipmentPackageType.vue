<template>
  <div v-if="query.data?.deliveryOptions.packageType">
    <PdkIcon :icon="query.data.deliveryOptions.packageType" />

    <span
      :title="translate('package_type')"
      v-text="translate(getPackageTypeTranslation(query.data.deliveryOptions.packageType))" />
  </div>
</template>

<script lang="ts" setup>
import {useQueryStore} from '../../stores';
import {getPackageTypeTranslation} from '../../helpers';
import {useLanguage} from '../../composables';

const props = defineProps<{shipmentId: number}>();

const query = useQueryStore().registerShipmentQuery(props.shipmentId);

const {translate} = useLanguage();
</script>
