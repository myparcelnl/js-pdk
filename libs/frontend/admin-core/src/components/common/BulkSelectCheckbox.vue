<template>
  <PdkCheckboxInput
    v-model="model"
    :element="bulkCheckboxElement"
    :title="translate('select_all')" />
</template>

<script lang="ts" setup>
import {computed} from 'vue';
import {Keyable} from '@myparcel-pdk/common';
import {createFormElement} from '../../utils';
import {useLanguage} from '../../composables';

const NAME_BULK = 'bulk';

const props = defineProps<{
  options: Keyable[];
  modelValue: Record<Keyable, boolean>;
}>();

const emit = defineEmits<(e: 'update:modelValue', value: Record<Keyable, boolean>) => void>();

const model = computed({
  get(): boolean {
    const values = Object.values(props.modelValue);

    return values.length > 0 && values.every(Boolean);
  },

  set(checked: boolean): void {
    emit('update:modelValue', Object.fromEntries(props.options.map((option) => [option, checked])));
  },
});

const bulkCheckboxElement = createFormElement(
  {
    name: NAME_BULK,
    ref: model,
    disabledWhen: () => props.options.length === 0,
  },
  NAME_BULK,
);

const {translate} = useLanguage();
</script>
