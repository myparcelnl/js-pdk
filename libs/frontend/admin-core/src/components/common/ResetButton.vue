<template>
  <PdkButton
    :disabled="disabled"
    :icon="AdminIcon.Refresh"
    label="action_reset"
    :variant="Variant.Secondary"
    @click="reset" />
</template>

<script lang="ts" setup>
import {Variant} from '@myparcel-pdk/common';
import {useForm} from '@myparcel/vue-form-builder';
import {AdminIcon} from '../../types';

defineProps<{
  disabled?: boolean;
}>();

const form = useForm();

const reset = () => {
  void form.reset();
};
</script>
