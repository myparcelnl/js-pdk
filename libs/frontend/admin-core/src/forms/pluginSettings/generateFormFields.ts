import {ref, toRaw} from 'vue';
import {get} from '@vueuse/core';
import {type Plugin} from '@myparcel-pdk/common';
import {
  type AnyElementConfiguration,
  type InteractiveElementConfiguration,
  defineField,
} from '@myparcel/vue-form-builder';
import {resolveFormComponent} from '../helpers';

type GenerateFormFields = (
  config: {
    fields: null | Plugin.Field[];
    values: Record<string, unknown>;
  },
  prefix?: string,
) => AnyElementConfiguration[];

export const generateFormFields: GenerateFormFields = ({fields, values}, prefix = '') => {
  if (!fields) {
    return [];
  }

  return fields.map((data) => {
    const {name, $component, $visibleWhen, $slot, $wrapper, label, ...props} = data;

    const common: AnyElementConfiguration = {
      component: resolveFormComponent($component),
      props: {...props},
      slots: $slot ? {default: $slot} : undefined,
      wrapper: $wrapper && typeof $wrapper === 'string' ? resolveFormComponent($wrapper) : undefined,
    };

    if ($visibleWhen) {
      common.visibleWhen = (field) => {
        return Object.entries($visibleWhen).every(([fieldName, value]) => {
          return get(field.form.model[prefix + fieldName]?.ref) === value;
        });
      };
    }

    // Plain element
    if (!label || !name) {
      return defineField(common);
    }

    return defineField({
      ...common,
      name: prefix + name,
      ref: ref(toRaw(values?.[name])),
      label,
    } as InteractiveElementConfiguration);
  });
};
