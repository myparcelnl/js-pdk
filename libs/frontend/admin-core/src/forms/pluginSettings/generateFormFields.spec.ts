import {describe, expect, it} from 'vitest';
import {AdminComponent, type Plugin} from '@myparcel-pdk/common';
import {type InteractiveElementConfiguration} from '@myparcel/vue-form-builder';
import {generateFormFields} from './generateFormFields';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const removeRef: any = (field: InteractiveElementConfiguration) => {
  const {ref, ...rest} = field;
  return rest;
};

describe('generateFormFields', () => {
  it('should generate fields from a view', () => {
    const view: Plugin.SettingsView = {
      id: 'general_settings',
      title: 'general_settings',
      description: 'general_settings_description',
      children: [],
      elements: [
        {
          name: 'generalSettings.name',
          label: 'general_settings_name',
          $component: AdminComponent.TextInput,
        },
        {
          name: 'generalSettings.description',
          label: 'general_settings_description',
          $component: AdminComponent.TextInput,
        },
      ],
    };

    const fields = generateFormFields({fields: view.elements, values: {}});
    expect(fields.map(removeRef)).toEqual([
      {
        component: 'PdkTextInput',
        label: 'general_settings_name',
        name: 'generalSettings.name',
        props: {},
      },
      {
        component: 'PdkTextInput',
        label: 'general_settings_description',
        name: 'generalSettings.description',
        props: {},
      },
    ]);
  });
});
