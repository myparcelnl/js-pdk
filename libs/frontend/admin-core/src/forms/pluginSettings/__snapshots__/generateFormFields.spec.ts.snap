// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateFormFields > should generate fields from a view 1`] = `
[
  {
    "attributes": {
      "max": 100,
      "min": 0,
      "step": 1,
    },
    "component": "PdkNumberInput",
    "label": "general_settings_name",
    "name": "generalSettings.name",
    "optional": true,
    "props": {
      "description": "general_settings_name_description",
      "subtext": "general_settings_name_subtext",
    },
    "slots": undefined,
    "wrapper": undefined,
  },
  {
    "attributes": {},
    "component": "PdkTextInput",
    "label": "general_settings_description",
    "name": "generalSettings.description",
    "optional": true,
    "props": {
      "description": "general_settings_description_description",
      "subtext": "general_settings_description_subtext",
    },
    "slots": undefined,
    "visibleWhen": [Function],
    "wrapper": undefined,
  },
]
`;
